allocations/tasks.py

```diff
@@ -1,18 +1,93 @@
-
-
-from datetime import datetime
+import calendar
+from datetime import datetime, timedelta
 import logging
 import pytz
 from operator import attrgetter
 
 from celery.decorators import task
+from django.conf import settings
+from django.core.mail import send_mail
 from django.db import transaction
+from django.utils.html import strip_tags
 from allocations.models import Allocation
 from allocations.allocations_api import BalanceServiceClient
 from util.keycloak_client import KeycloakClient
 
 LOG = logging.getLogger(__name__)
 
+def warn_user_for_expiring_allocation():
+    """
+    Sends an email to users when their allocation is within one month of expiring
+    """
+    now = datetime.now()
+    today = now.date()
+    days_this_month = calendar.monthrange(today.year, today.month)[1]
+    target_expiration_date = today + timedelta(days=days_this_month)
+
+    # Find allocations that are expiring between today and 1 month, and have not been warned of their impending doom
+    expiring_allocations = Allocation.objects.filter(
+        status='active',
+        expiration_date__lte=target_expiration_date,
+        expiration_date__gte=today,
+        expiration_warning_issued__isnull=True
+    )
+    emails_sent = 0
+    for alloc in expiring_allocations:
+        charge_code = alloc.project.charge_code
+        expiration_date = alloc.expiration_date.date()
+
+        if not expiration_date:
+            LOG.debug(f"Allocation {alloc.id} has no expiration date.")
+            continue
+
+        time_until_expiration = expiration_date - today
+
+        if expiration_date == today:
+            time_description = "today"
+        else:
+            # "in 1 day" | "in <n> days"
+            time_description = f"in {time_until_expiration.days} day{'s' if time_until_expiration.days != 1 else ''}"
+
+        email_body = f"""
+        <p>
+            The allocation for project <b>{charge_code}</b> will expire <b>{time_description}.</b>
+            See our
+            <a href="https://chameleoncloud.readthedocs.io/en/latest/user/project.html#recharge-or-extend-your-allocation">Documentation</a>
+            on how to recharge or extend your allocation.
+        </p>
+        """
+
+        mail_error = f"Failed to send expiration warning email for project {charge_code} to {alloc.project.pi.email}"
+        try:
+            mail_sent = send_mail(
+                subject=f"NOTICE: Your allocation for project {charge_code} expires {time_description}!",
+                from_email=settings.DEFAULT_FROM_EMAIL,
+                recipient_list=[alloc.project.pi.email],
```

Comment:
IIRC there are possibly a few users who don't have the email attribute set--I think we should check for that here and issue a warning if so.

Reply:
Yes, I think you can set the configuration in settings.py to set the interval to something like 24 hours, or maybe on a calendar cron-like syntax?

https://stackoverflow.com/questions/32449845/how-to-run-a-django-celery-task-every-6am-and-6pm-daily

Reply:
I changed the interval. I think if it's only run once per day, it's ok if we have repeat warnings in our logs about projects with no PI email. How do you feel about that?

Reply:
Yep, sounds good to me :)